### PR TITLE
feat!: cosmwasm an optional integration to app

### PR DESCRIPTION
**Pull request by gsk967**
## Description

added cosmwasm feature as an optional features for app 

```bash
// to build cosmwasm feature into app 
$ EXPERIMENTAL=true make build
```

closes: #802 

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -5,6 +5,7 @@ go 1.19
 require (
 	cosmossdk.io/errors v1.0.0-beta.7
 	cosmossdk.io/math v1.0.0-beta.3
+	github.com/CosmWasm/wasmd v0.29.0-pio-1
```
Comment (robert-zaremba) [inline go.mod:8]: can we use `v0.29.1`?

File: go.mod [hunk 2]
```diff
@@ -324,6 +327,7 @@ require (
 )
 
 replace (
+	github.com/CosmWasm/wasmd => github.com/provenance-io/wasmd v0.29.0-pio-1
```
Comment (robert-zaremba) [inline go.mod:330]: let's use notional fork, it's more up to date: https://github.com/notional-labs/wasmd/tree/v0.29.0-sdk46

File: Makefile [hunk 3]
```diff
@@ -30,6 +30,11 @@ endif
 
 build_tags = netgo
```
Comment (robert-zaremba) [inline Makefile:32]: let's add next to the build job:
```
build-experimental: go.sum
	@echo "--> Building Experimental version..."
	EXPERIMENTAL=true go build -mod=readonly -o $(BUILD_DIR)/ $(BUILD_FLAGS) ./...
```

File: app/experimental_appconfig.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+	proposals, err := wasm.ConvertToProposals(chunks)
+	if err != nil {
+		panic(err)
+	}
+
+	return proposals
+}
+
+func setCustomModuleBasics() []module.AppModuleBasic {
```
Comment (robert-zaremba) [inline app/experimental_appconfig.go:77]:
```suggestion
func customModuleBasics() []module.AppModuleBasic {
```

Comment (robert-zaremba) [reply app/experimental_appconfig.go:77]: this function is not setting anything. Let's update the names below as well.

File: app/experimental_appconfig.go [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+func setCustomOrderBeginBlocker() []string {
+	return []string{
+		wasm.ModuleName,
+	}
+}
+
+func setCustomOrderEndBlocker() []string {
+	return []string{
+		wasm.ModuleName,
+	}
```
Comment (robert-zaremba) [inline app/experimental_appconfig.go:113]:
```suggestion
	return []string{wasm.ModuleName}
```

Comment (robert-zaremba) [reply app/experimental_appconfig.go:113]: maybe we can apply same thing above?

File: app/stable_appconfig.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+	return []string{}
+}
+
+func setCustomOrderMigrations() []string {
+	return []string{}
+}
+
+func initCustomParamsKeeper(_ *paramskeeper.Keeper) {}
+
+func setCustomProposalHanndlers() []govclient.ProposalHandler {
```
Comment (robert-zaremba) [inline app/stable_appconfig.go:73]:
```suggestion
func customProposalHanndlers() []govclient.ProposalHandler {
```

File: app/stable_appconfig.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+		distrclient.ProposalHandler,
+		upgradeclient.LegacyProposalHandler,
+		upgradeclient.LegacyCancelProposalHandler,
+		leverageclient.ProposalHandler,
+		ibcclientclient.UpdateClientProposalHandler,
+		ibcclientclient.UpgradeProposalHandler,
+	}
+}
+
+func (app *UmeeApp) setCustomAnteHandler(txConfig client.TxConfig,
```
Comment (robert-zaremba) [inline app/stable_appconfig.go:85]:
```suggestion
func (app *UmeeApp) customAnteHandler(txConfig client.TxConfig,
```

**Review by robert-zaremba — COMMENTED** [review]
Left few comments.
Also, have an idea how to better integrate the experimental build without doing so many custom functions. I will make a commit with an example directly in your repo, let me know what do you think.

**Comment by robert-zaremba** [comment]
See this commit as an alternative idea to reduce copied functions: 009f87d
What do you think?

File: .goreleaser.yml [hunk 8]
```diff
@@ -3,6 +3,7 @@ project_name: umee
 
 env:
   - GO111MODULE=on
+  - CGO_ENABLED=1
```
[the remark this answers is not in the document]
Comment (gsk967) [reply .goreleaser.yml:6]: We need this because `cosmvm` is doesn't compiling when `CGO_ENABLED=0`

**Comment by gsk967** [comment]
> See this commit as an alternative idea to reduce copied functions: [009f87d](https://github.com/umee-network/umee/commit/009f87d29478acdccaa6e51b49ff3dbfb813d9d6) What do you think?

I think it is better to keep old functions only, because we already passing one build flag for `experimental config` , so no need for new flag again and for we will keep experimental features everything into `experimenatal_appconfig`  only.

**Comment by robert-zaremba** [comment]
> I think it is better to keep old functions only, because we already passing one build flag for experimental config , so no need for new flag again.

There is no need for a new flag. Where did you get an impression that we need a new flag?

> We will keep experimental features everything into experimenatal_appconfig only.

On the other hand, the code is a bit more bloated (harder to follow).

File: .goreleaser.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply .goreleaser.yml:6]: OK, please add a comment in the file.

**Comment by gsk967** [comment]
@robert-zaremba  Please can you review again

File: umee.e2e.Dockerfile [hunk 9]
```diff
@@ -13,22 +13,31 @@ FROM umee-base-builder AS umeed-builder
...
-ARG PEGGO_VERSION=v0.3.0
-WORKDIR /downloads/
-RUN git clone https://github.com/umee-network/peggo.git
-RUN cd peggo && git checkout ${PEGGO_VERSION} && make build && cp ./build/peggo /usr/local/bin/
+# # Fetch peggo (gravity bridge) binary
+# FROM base-builder AS peggo-builder
+# ARG PEGGO_VERSION=v0.3.0
+# WORKDIR /downloads/
+# RUN git clone https://github.com/umee-network/peggo.git
+# RUN cd peggo && git checkout ${PEGGO_VERSION} && make build && cp ./build/peggo /usr/local/bin/
```
Comment (robert-zaremba) [inline umee.e2e.Dockerfile:34]: why this is commented?

**Review by robert-zaremba — COMMENTED** [review]
Looks good, let's solve the remaining comments.

**Review by robert-zaremba — APPROVED** [review]

**Pull request merged**